File: webapp/src/libs/models/Specialization.ts
export interface ISpecialization {
    id: string;
    label: string;
    name: string;
    description: string;
    roleInformation: string;
    indexId: string;
    openAIDeploymentId: string;
    completionDeploymentName: string;
    imageFilePath: string;
    iconFilePath: string;
    isActive: boolean;
    groupMemberships: string[];
    isDefault: boolean;
    restrictResultScope: boolean | null;
    strictness: number | null;
    documentCount: number | null;
    pastMessagesIncludedCount: number | null;
    maxResponseTokenLimit: number | null;
    initialChatMessage: string;
    order: number;
    suggestions: string[];
    canGenImages: boolean;
    enableKernelMemoryMultiIndex: boolean;
    indexIds?: string[];
}

/**
 * Specialization request interface.
 *
 */
export interface ISpecializationRequest {
    label: string;
    name: string;
    description: string;
    roleInformation: string;
    indexId: string;
    openAIDeploymentId: string;
    completionDeploymentName: string;
    groupMemberships: string[];
    initialChatMessage: string;
    isDefault: boolean;
    restrictResultScope: boolean | null;
    strictness: number | null;
    documentCount: number | null;
    pastMessagesIncludedCount: number | null;
    maxResponseTokenLimit: number | null;
    order: number;
    suggestions: string[];
    canGenImages: boolean;
    enableKernelMemoryMultiIndex: boolean;
    indexIds?: string[];
}

export interface ISpecializationToggleRequest {
    isActive: boolean;
}

export interface IChatCompletionDeployment {
    name: string;
    completionTokenLimit: number;
    outputTokens: number;
}

File: webapi/Models/Storage/Specialization.cs
// Copyright (c) Quartech. All rights reserved.

using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using CopilotChat.WebApi.Storage;

namespace CopilotChat.WebApi.Models.Storage;

public record Specialization : IStorageEntity
{
    public string Id { get; set; }

    public string Label { get; set; }

    public string Name { get; set; }

    public string Description { get; set; }

    public string InitialChatMessage { get; set; }

    public string RoleInformation { get; set; }

    public IList<string> GroupMemberships { get; set; } = new List<string>();

    public string? IndexId { get; set; }

    public string OpenAIDeploymentId { get; set; }

    public string? CompletionDeploymentName { get; set; }

    public string ImageFilePath { get; set; }

    public string IconFilePath { get; set; }

    [JsonIgnore]
    public string Partition => this.Id;

    public bool IsActive { get; set; }

    public string CreatedBy { get; set; } = "";

    public string UpdatedBy { get; set; } = "";

    public DateTimeOffset CreatedOn { get; set; }

    public DateTimeOffset UpdatedAt { get; set; }

    public bool? IsDefault { get; set; }

    public bool? RestrictResultScope { get; set; }

    public int? Strictness { get; set; }

    public int? DocumentCount { get; set; }

    public int? PastMessagesIncludedCount { get; set; }

    public int? MaxResponseTokenLimit { get; set; }

    public int? Order { get; set; }

    public IList<string> Suggestions { get; set; } = new List<string>();

    public bool CanGenImages { get; set; } = false;
}
